lib/rules/no-useless-assertions.ts

```diff
@@ -273,24 +397,60 @@ export default createRule("no-useless-assertions", {
                 }
             }
 
+            /**
+             * Verify for Assertion
+             */
+            function verifyAssertion(
+                assertion: Assertion,
+                getFirstCharAfterFn: GetFirstCharAfter,
+            ): void {
+                switch (assertion.kind) {
+                    case "start":
+                    case "end":
+                        verifyStartOrEnd(assertion, getFirstCharAfterFn)
+                        break
+
+                    case "word":
+                        verifyWordBoundary(assertion, getFirstCharAfterFn)
+                        break
+
+                    case "lookahead":
+                    case "lookbehind":
+                        verifyLookaround(assertion, getFirstCharAfterFn)
+                        break
+                    default:
+                }
+            }
+
             return {
-                onAssertionEnter(assertion) {
-                    switch (assertion.kind) {
-                        case "start":
-                        case "end":
-                            verifyStartOrEnd(assertion)
-                            break
-
-                        case "word":
-                            verifyWordBoundary(assertion)
-                            break
-
-                        case "lookahead":
-                        case "lookbehind":
-                            verifyLookaround(assertion)
-                            break
-                        default:
-                    }
+                onPatternEnter(pattern) {
+                    // Phase 1:
+                    // The context of assertions is determined by only looking
+                    // at elements after the current assertion. This means that
+                    // the order of assertions is kept as is.
+                    visitRegExpAST(pattern, {
+                        onAssertionEnter(assertion) {
+                            verifyAssertion(assertion, getFirstCharAfter)
+                        },
+                    })
+
+                    // Phase 2:
+                    // The context of assertions is determined by reordering
+                    // assertions such that as much information as possible can
+                    // be extracted from its surrounding assertions.
+                    const reorderingGetFirstCharAfter =
+                        createReorderingGetFirstCharAfter(reported)
+                    visitRegExpAST(pattern, {
+                        onAssertionEnter(assertion) {
+                            if (reported.has(assertion)) {
+                                return
+                            }
+                            verifyAssertion(
+                                assertion,
+                                reorderingGetFirstCharAfter,
+                            )
+                        },
+                    })
```

Comment:
By collecting the nodes visited in the traverse, I think it will be unnecessary to trace from the root in the second traverse.


```ts

            const allAssertions: Assertion[] = []
            return {
                onAssertionEnter(assertion) {
                    // Phase 1:
                    // ...
                    verifyAssertion(assertion, getFirstCharAfter)
                    allAssertions.push(assertion)
                },
                onPatternLeave() {
                    // Phase 2:
                    // ...
                    const reorderingGetFirstCharAfter =
                        createReorderingGetFirstCharAfter(reported)
                    for (const assertion of allAssertions) {
                        if (reported.has(assertion)) {
                            continue
                        }
                        verifyAssertion(assertion, reorderingGetFirstCharAfter)
                    }
                },
            }
```

Reply:
Good point!